// @livekit/protocol/src/gen/livekit_agent_simulation_pb.js
// Copyright 2024 LiveKit, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @generated by protoc-gen-es v1.10.1 with parameter "target=dts+js"
// @generated from file livekit_agent_simulation.proto (package livekit, syntax proto3)
/* eslint-disable */
// @ts-nocheck

import { proto3, Timestamp } from "@bufbuild/protobuf";
import { ChatContext } from "./agent/livekit_agent_session_pb.js";
import { PresignedPostRequest } from "./livekit_cloud_agent_pb.js";
import { TokenPagination } from "./livekit_models_pb.js";

/**
 * @generated from message livekit.SimulationRunSummary
 */
export const SimulationRunSummary = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRunSummary",
  () => [
    { no: 1, name: "passed", kind: "scalar", T: 5 /* ScalarType.INT32 */ },
    { no: 2, name: "failed", kind: "scalar", T: 5 /* ScalarType.INT32 */ },
    { no: 3, name: "going_well", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 4, name: "to_improve", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 5, name: "issues", kind: "message", T: SimulationRunSummary_Issue, repeated: true },
    { no: 6, name: "chat_history", kind: "map", K: 9 /* ScalarType.STRING */, V: {kind: "message", T: ChatContext} },
  ],
);

/**
 * @generated from message livekit.SimulationRunSummary.Issue
 */
export const SimulationRunSummary_Issue = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRunSummary.Issue",
  () => [
    { no: 1, name: "description", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "suggestion", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
  {localName: "SimulationRunSummary_Issue"},
);

/**
 * @generated from message livekit.SimulationRun
 */
export const SimulationRun = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun",
  () => [
    { no: 1, name: "id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "status", kind: "enum", T: proto3.getEnumType(SimulationRun_Status) },
    { no: 4, name: "agent_description", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 5, name: "error", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 6, name: "created_at", kind: "message", T: Timestamp },
    { no: 7, name: "jobs", kind: "message", T: SimulationRun_Job, repeated: true },
    { no: 8, name: "summary", kind: "message", T: SimulationRunSummary },
  ],
);

/**
 * @generated from enum livekit.SimulationRun.Status
 */
export const SimulationRun_Status = /*@__PURE__*/ proto3.makeEnum(
  "livekit.SimulationRun.Status",
  [
    {no: 0, name: "STATUS_PENDING_UPLOAD", localName: "PENDING_UPLOAD"},
    {no: 1, name: "STATUS_GENERATING", localName: "GENERATING"},
    {no: 2, name: "STATUS_RUNNING", localName: "RUNNING"},
    {no: 3, name: "STATUS_SUMMARIZING", localName: "SUMMARIZING"},
    {no: 4, name: "STATUS_COMPLETED", localName: "COMPLETED"},
    {no: 5, name: "STATUS_FAILED", localName: "FAILED"},
    {no: 6, name: "STATUS_CANCELLED", localName: "CANCELLED"},
  ],
);

/**
 * @generated from message livekit.SimulationRun.Job
 */
export const SimulationRun_Job = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.Job",
  () => [
    { no: 1, name: "id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "status", kind: "enum", T: proto3.getEnumType(SimulationRun_Job_Status) },
    { no: 3, name: "instructions", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 4, name: "error", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 5, name: "agent_expectations", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 6, name: "label", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
  {localName: "SimulationRun_Job"},
);

/**
 * @generated from enum livekit.SimulationRun.Job.Status
 */
export const SimulationRun_Job_Status = /*@__PURE__*/ proto3.makeEnum(
  "livekit.SimulationRun.Job.Status",
  [
    {no: 0, name: "STATUS_PENDING", localName: "PENDING"},
    {no: 1, name: "STATUS_RUNNING", localName: "RUNNING"},
    {no: 2, name: "STATUS_COMPLETED", localName: "COMPLETED"},
    {no: 3, name: "STATUS_FAILED", localName: "FAILED"},
    {no: 4, name: "STATUS_CANCELLED", localName: "CANCELLED"},
  ],
);

/**
 * @generated from message livekit.SimulationRun.Create
 */
export const SimulationRun_Create = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.Create",
  [],
  {localName: "SimulationRun_Create"},
);

/**
 * @generated from message livekit.SimulationRun.Create.Scenario
 */
export const SimulationRun_Create_Scenario = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.Create.Scenario",
  () => [
    { no: 1, name: "instructions", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "agent_expectations", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "metadata", kind: "map", K: 9 /* ScalarType.STRING */, V: {kind: "scalar", T: 9 /* ScalarType.STRING */} },
    { no: 4, name: "label", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
  {localName: "SimulationRun_Create_Scenario"},
);

/**
 * @generated from message livekit.SimulationRun.Create.Scenarios
 */
export const SimulationRun_Create_Scenarios = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.Create.Scenarios",
  () => [
    { no: 1, name: "scenarios", kind: "message", T: SimulationRun_Create_Scenario, repeated: true },
  ],
  {localName: "SimulationRun_Create_Scenarios"},
);

/**
 * @generated from message livekit.SimulationRun.Create.Request
 */
export const SimulationRun_Create_Request = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.Create.Request",
  () => [
    { no: 1, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "agent_name", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "agent_description", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 4, name: "num_simulations", kind: "scalar", T: 5 /* ScalarType.INT32 */ },
    { no: 6, name: "region", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 7, name: "scenarios", kind: "message", T: SimulationRun_Create_Scenarios, oneof: "source" },
    { no: 5, name: "group_id", kind: "scalar", T: 9 /* ScalarType.STRING */, oneof: "source" },
  ],
  {localName: "SimulationRun_Create_Request"},
);

/**
 * @generated from message livekit.SimulationRun.Create.Response
 */
export const SimulationRun_Create_Response = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.Create.Response",
  () => [
    { no: 1, name: "simulation_run_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "presigned_post_request", kind: "message", T: PresignedPostRequest },
  ],
  {localName: "SimulationRun_Create_Response"},
);

/**
 * @generated from message livekit.SimulationRun.ConfirmSourceUpload
 */
export const SimulationRun_ConfirmSourceUpload = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.ConfirmSourceUpload",
  [],
  {localName: "SimulationRun_ConfirmSourceUpload"},
);

/**
 * @generated from message livekit.SimulationRun.ConfirmSourceUpload.Request
 */
export const SimulationRun_ConfirmSourceUpload_Request = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.ConfirmSourceUpload.Request",
  () => [
    { no: 1, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "simulation_run_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "code_entrypoint", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
  {localName: "SimulationRun_ConfirmSourceUpload_Request"},
);

/**
 * @generated from message livekit.SimulationRun.ConfirmSourceUpload.Response
 */
export const SimulationRun_ConfirmSourceUpload_Response = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.ConfirmSourceUpload.Response",
  [],
  {localName: "SimulationRun_ConfirmSourceUpload_Response"},
);

/**
 * @generated from message livekit.SimulationRun.Get
 */
export const SimulationRun_Get = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.Get",
  [],
  {localName: "SimulationRun_Get"},
);

/**
 * @generated from message livekit.SimulationRun.Get.Request
 */
export const SimulationRun_Get_Request = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.Get.Request",
  () => [
    { no: 1, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "simulation_run_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
  {localName: "SimulationRun_Get_Request"},
);

/**
 * @generated from message livekit.SimulationRun.Get.Response
 */
export const SimulationRun_Get_Response = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.Get.Response",
  () => [
    { no: 1, name: "run", kind: "message", T: SimulationRun },
  ],
  {localName: "SimulationRun_Get_Response"},
);

/**
 * @generated from message livekit.SimulationRun.List
 */
export const SimulationRun_List = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.List",
  [],
  {localName: "SimulationRun_List"},
);

/**
 * @generated from message livekit.SimulationRun.List.Request
 */
export const SimulationRun_List_Request = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.List.Request",
  () => [
    { no: 1, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "status", kind: "enum", T: proto3.getEnumType(SimulationRun_Status), opt: true },
    { no: 3, name: "page_token", kind: "message", T: TokenPagination, opt: true },
  ],
  {localName: "SimulationRun_List_Request"},
);

/**
 * @generated from message livekit.SimulationRun.List.Response
 */
export const SimulationRun_List_Response = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.List.Response",
  () => [
    { no: 1, name: "runs", kind: "message", T: SimulationRun, repeated: true },
    { no: 2, name: "next_page_token", kind: "message", T: TokenPagination },
  ],
  {localName: "SimulationRun_List_Response"},
);

/**
 * @generated from message livekit.SimulationRun.Cancel
 */
export const SimulationRun_Cancel = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.Cancel",
  [],
  {localName: "SimulationRun_Cancel"},
);

/**
 * @generated from message livekit.SimulationRun.Cancel.Request
 */
export const SimulationRun_Cancel_Request = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.Cancel.Request",
  () => [
    { no: 1, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "simulation_run_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
  {localName: "SimulationRun_Cancel_Request"},
);

/**
 * @generated from message livekit.SimulationRun.Cancel.Response
 */
export const SimulationRun_Cancel_Response = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.SimulationRun.Cancel.Response",
  [],
  {localName: "SimulationRun_Cancel_Response"},
);

/**
 * @generated from message livekit.Scenario
 */
export const Scenario = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario",
  () => [
    { no: 1, name: "id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "label", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 4, name: "instructions", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 5, name: "agent_expectations", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 6, name: "source_room_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 7, name: "created_at", kind: "message", T: Timestamp },
    { no: 8, name: "status", kind: "enum", T: proto3.getEnumType(Scenario_Status) },
    { no: 9, name: "error", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 10, name: "group_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 11, name: "metadata", kind: "map", K: 9 /* ScalarType.STRING */, V: {kind: "scalar", T: 9 /* ScalarType.STRING */} },
  ],
);

/**
 * @generated from enum livekit.Scenario.Status
 */
export const Scenario_Status = /*@__PURE__*/ proto3.makeEnum(
  "livekit.Scenario.Status",
  [
    {no: 0, name: "STATUS_GENERATING", localName: "GENERATING"},
    {no: 1, name: "STATUS_READY", localName: "READY"},
    {no: 2, name: "STATUS_FAILED", localName: "FAILED"},
  ],
);

/**
 * @generated from message livekit.Scenario.Create
 */
export const Scenario_Create = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.Create",
  [],
  {localName: "Scenario_Create"},
);

/**
 * @generated from message livekit.Scenario.Create.Request
 */
export const Scenario_Create_Request = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.Create.Request",
  () => [
    { no: 1, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "label", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "instructions", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 4, name: "agent_expectations", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 5, name: "group_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 6, name: "metadata", kind: "map", K: 9 /* ScalarType.STRING */, V: {kind: "scalar", T: 9 /* ScalarType.STRING */} },
  ],
  {localName: "Scenario_Create_Request"},
);

/**
 * @generated from message livekit.Scenario.Create.Response
 */
export const Scenario_Create_Response = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.Create.Response",
  () => [
    { no: 1, name: "scenario", kind: "message", T: Scenario },
  ],
  {localName: "Scenario_Create_Response"},
);

/**
 * @generated from message livekit.Scenario.CreateFromSession
 */
export const Scenario_CreateFromSession = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.CreateFromSession",
  [],
  {localName: "Scenario_CreateFromSession"},
);

/**
 * @generated from message livekit.Scenario.CreateFromSession.Request
 */
export const Scenario_CreateFromSession_Request = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.CreateFromSession.Request",
  () => [
    { no: 1, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "room_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "group_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 4, name: "label", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 5, name: "region", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
  {localName: "Scenario_CreateFromSession_Request"},
);

/**
 * @generated from message livekit.Scenario.CreateFromSession.Response
 */
export const Scenario_CreateFromSession_Response = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.CreateFromSession.Response",
  () => [
    { no: 1, name: "scenario", kind: "message", T: Scenario },
  ],
  {localName: "Scenario_CreateFromSession_Response"},
);

/**
 * @generated from message livekit.Scenario.Delete
 */
export const Scenario_Delete = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.Delete",
  [],
  {localName: "Scenario_Delete"},
);

/**
 * @generated from message livekit.Scenario.Delete.Request
 */
export const Scenario_Delete_Request = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.Delete.Request",
  () => [
    { no: 1, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "scenario_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
  {localName: "Scenario_Delete_Request"},
);

/**
 * @generated from message livekit.Scenario.Delete.Response
 */
export const Scenario_Delete_Response = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.Delete.Response",
  [],
  {localName: "Scenario_Delete_Response"},
);

/**
 * @generated from message livekit.Scenario.Update
 */
export const Scenario_Update = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.Update",
  [],
  {localName: "Scenario_Update"},
);

/**
 * @generated from message livekit.Scenario.Update.Request
 */
export const Scenario_Update_Request = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.Update.Request",
  () => [
    { no: 1, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "scenario_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "label", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 4, name: "instructions", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 5, name: "agent_expectations", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 6, name: "group_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 7, name: "metadata", kind: "map", K: 9 /* ScalarType.STRING */, V: {kind: "scalar", T: 9 /* ScalarType.STRING */} },
  ],
  {localName: "Scenario_Update_Request"},
);

/**
 * @generated from message livekit.Scenario.Update.Response
 */
export const Scenario_Update_Response = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.Update.Response",
  () => [
    { no: 1, name: "scenario", kind: "message", T: Scenario },
  ],
  {localName: "Scenario_Update_Response"},
);

/**
 * @generated from message livekit.Scenario.List
 */
export const Scenario_List = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.List",
  [],
  {localName: "Scenario_List"},
);

/**
 * @generated from message livekit.Scenario.List.Request
 */
export const Scenario_List_Request = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.List.Request",
  () => [
    { no: 1, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "group_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
  {localName: "Scenario_List_Request"},
);

/**
 * @generated from message livekit.Scenario.List.Response
 */
export const Scenario_List_Response = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.Scenario.List.Response",
  () => [
    { no: 1, name: "scenarios", kind: "message", T: Scenario, repeated: true },
  ],
  {localName: "Scenario_List_Response"},
);

/**
 * @generated from message livekit.ScenarioGroup
 */
export const ScenarioGroup = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.ScenarioGroup",
  () => [
    { no: 1, name: "id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "label", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 4, name: "created_at", kind: "message", T: Timestamp },
    { no: 5, name: "scenarios", kind: "message", T: Scenario, repeated: true },
  ],
);

/**
 * @generated from message livekit.ScenarioGroup.Create
 */
export const ScenarioGroup_Create = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.ScenarioGroup.Create",
  [],
  {localName: "ScenarioGroup_Create"},
);

/**
 * @generated from message livekit.ScenarioGroup.Create.Request
 */
export const ScenarioGroup_Create_Request = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.ScenarioGroup.Create.Request",
  () => [
    { no: 1, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "label", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
  {localName: "ScenarioGroup_Create_Request"},
);

/**
 * @generated from message livekit.ScenarioGroup.Create.Response
 */
export const ScenarioGroup_Create_Response = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.ScenarioGroup.Create.Response",
  () => [
    { no: 1, name: "scenario_group", kind: "message", T: ScenarioGroup },
  ],
  {localName: "ScenarioGroup_Create_Response"},
);

/**
 * @generated from message livekit.ScenarioGroup.Delete
 */
export const ScenarioGroup_Delete = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.ScenarioGroup.Delete",
  [],
  {localName: "ScenarioGroup_Delete"},
);

/**
 * @generated from message livekit.ScenarioGroup.Delete.Request
 */
export const ScenarioGroup_Delete_Request = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.ScenarioGroup.Delete.Request",
  () => [
    { no: 1, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "group_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
  {localName: "ScenarioGroup_Delete_Request"},
);

/**
 * @generated from message livekit.ScenarioGroup.Delete.Response
 */
export const ScenarioGroup_Delete_Response = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.ScenarioGroup.Delete.Response",
  [],
  {localName: "ScenarioGroup_Delete_Response"},
);

/**
 * @generated from message livekit.ScenarioGroup.List
 */
export const ScenarioGroup_List = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.ScenarioGroup.List",
  [],
  {localName: "ScenarioGroup_List"},
);

/**
 * @generated from message livekit.ScenarioGroup.List.Request
 */
export const ScenarioGroup_List_Request = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.ScenarioGroup.List.Request",
  () => [
    { no: 1, name: "project_id", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ],
  {localName: "ScenarioGroup_List_Request"},
);

/**
 * @generated from message livekit.ScenarioGroup.List.Response
 */
export const ScenarioGroup_List_Response = /*@__PURE__*/ proto3.makeMessageType(
  "livekit.ScenarioGroup.List.Response",
  () => [
    { no: 1, name: "scenario_groups", kind: "message", T: ScenarioGroup, repeated: true },
  ],
  {localName: "ScenarioGroup_List_Response"},
);
