// @livekit/components-react/dist/hooks.mjs
import { D as e, M as t, S as i, E as u, F as n, G as o, H as r, w as c, b as p, u as P, i as d, a as l, I as T, c as g, J as S, K as m, d as k, n as v, m as A, L as C, N as I, O as M, C as R, A as b, e as f, y, s as D, o as E, P as V, Q as h, R as L, T as x, l as B, U, V as F, x as w, B as G, W as K, X as N, v as Q, Y as W, Z as q, _ as H, $ as J, a0 as O, a1 as X, f as Y, g as Z, p as _, a2 as $, a3 as j, a4 as z, k as ss, h as as, a5 as es, a6 as ts, t as is, a7 as us, r as ns, a8 as os } from "./hooks-CA4A5LBF.mjs";
import { u as cs } from "./room-Q0GUQkyk.mjs";
export {
  e as AgentEvent,
  t as MessagesEvent,
  i as SessionEvent,
  u as isUseSessionReturn,
  n as useAgent,
  o as useAudioPlayback,
  r as useAudioWaveform,
  c as useChat,
  p as useChatToggle,
  P as useClearPinButton,
  d as useConnectionQualityIndicator,
  l as useConnectionState,
  T as useDataChannel,
  g as useDisconnectButton,
  S as useEvents,
  m as useFacingMode,
  k as useFocusToggle,
  v as useGridLayout,
  A as useIsEncrypted,
  C as useIsMuted,
  I as useIsRecording,
  M as useIsSpeaking,
  cs as useLiveKitRoom,
  R as useLocalParticipant,
  b as useLocalParticipantPermissions,
  f as useMediaDeviceSelect,
  y as useMediaDevices,
  D as useMultibandTrackVolume,
  E as usePagination,
  V as useParticipantAttribute,
  h as useParticipantAttributes,
  L as useParticipantInfo,
  x as useParticipantPermissions,
  B as useParticipantTile,
  U as useParticipantTracks,
  F as useParticipants,
  w as usePersistentUserChoices,
  G as usePinnedTracks,
  K as useRemoteParticipant,
  N as useRemoteParticipants,
  Q as useRoomInfo,
  W as useRpc,
  q as useSequentialRoomConnectDisconnect,
  H as useSession,
  J as useSessionMessages,
  O as useSortedParticipants,
  X as useSpeakingParticipants,
  Y as useStartAudio,
  Z as useStartVideo,
  _ as useSwipe,
  $ as useTextStream,
  j as useToken,
  z as useTrackByName,
  ss as useTrackMutedIndicator,
  as as useTrackToggle,
  es as useTrackTranscription,
  ts as useTrackVolume,
  is as useTracks,
  us as useTranscriptions,
  ns as useVisualStableUpdate,
  os as useVoiceAssistant
};
