// @livekit/components-react/dist/krisp.mjs
import * as t from "react";
import { LocalAudioTrack as b } from "livekit-client";
import { l as c } from "./contexts-DzwYztG5.mjs";
import { C as P } from "./hooks-CA4A5LBF.mjs";
function N(l = {}) {
  const [i, f] = t.useState(!1), [u, o] = t.useState(!1), [p, n] = t.useState(!1);
  let r = P().microphoneTrack;
  const [s, d] = t.useState();
  l.trackRef && (r = l.trackRef.publication);
  const m = t.useCallback(async (e) => {
    if (e) {
      const { KrispNoiseFilter: a, isKrispNoiseFilterSupported: k } = await import("@livekit/krisp-noise-filter");
      if (!k()) {
        c.warn("LiveKit-Krisp noise filter is not supported in this browser");
        return;
      }
      s || d(a(l.filterOptions));
    }
    f((a) => (a !== e && o(!0), e));
  }, []);
  return t.useEffect(() => {
    if (r && r.track instanceof b && s) {
      const e = r.track.getProcessor();
      e && e.name === "livekit-noise-filter" ? (o(!0), e.setEnabled(i).finally(() => {
        o(!1), n(i);
      })) : !e && i && (o(!0), r?.track?.setProcessor(s).then(() => s.setEnabled(i)).then(() => {
        n(!0);
      }).catch((a) => {
        n(!1), c.error("Krisp hook: error enabling filter", a);
      }).finally(() => {
        o(!1);
      }));
    }
  }, [i, r, s]), {
    setNoiseFilterEnabled: m,
    isNoiseFilterEnabled: p,
    isNoiseFilterPending: u,
    processor: s
  };
}
export {
  N as useKrispNoiseFilter
};
